cdap-app-fabric/src/main/java/co/cask/cdap/metadata/DefaultMetadataAdmin.java

```diff
@@ -164,5 +191,111 @@ private void ensureEntityExists(Id.NamespacedId entityId) throws NamespaceNotFou
     } catch (Exception e) {
       throw Throwables.propagate(e);
     }
+
+    // Check existence of entity
+    String targetType = getTargetType(entityId);
+    if (Id.Program.class.getSimpleName().equals(targetType)) {
+      Id.Program program = (Id.Program) entityId;
+      Id.Application application = program.getApplication();
+      ApplicationSpecification appSpec = store.getApplication(application);
+      if (appSpec == null) {
+        throw new ApplicationNotFoundException(application);
+      }
+      ensureProgramExists(appSpec, program);
+    } else if (Id.Application.class.getSimpleName().equals(targetType)) {
+      Id.Application application = (Id.Application) entityId;
+      if (store.getApplication(application) == null) {
+        throw new ApplicationNotFoundException(application);
+      }
+    } else if (Id.DatasetInstance.class.getSimpleName().equals(targetType)) {
+      Id.DatasetInstance datasetInstance = (Id.DatasetInstance) entityId;
+      try {
+        if (!datasetFramework.hasInstance(datasetInstance)) {
+          throw new DatasetNotFoundException(datasetInstance);
+        }
+      } catch (DatasetManagementException | ServiceUnavailableException ex) {
+        throw new DatasetNotFoundException(datasetInstance);
+      }
+    } else if (Id.Stream.class.getSimpleName().equals(targetType)) {
+      Id.Stream stream = (Id.Stream) entityId;
+      try {
+        if (!streamAdmin.exists(stream)) {
+          throw new StreamNotFoundException(stream);
+        }
+      } catch (Exception ex)  {
+        throw new StreamNotFoundException(stream);
```

Comment:
It was already throwing not found exception due to check for missing namespace. Lets address it in different PR